src/hotspot/share/prims/jvmtiTagMap.cpp

```diff
@@ -2637,120 +2776,83 @@ inline bool VM_HeapWalkOperation::collect_simple_roots() {
   return true;
 }
 
-// Walk the stack of a given thread and find all references (locals
-// and JNI calls) and report these as stack references
+// Reports the thread as JVMTI_HEAP_REFERENCE_THREAD,
+// walks the stack of the thread, finds all references (locals
+// and JNI calls) and reports these as stack references
 inline bool VM_HeapWalkOperation::collect_stack_roots(JavaThread* java_thread,
                                                       JNILocalRootsClosure* blk)
 {
   oop threadObj = java_thread->threadObj();
+  oop mounted_vt = java_thread->is_vthread_mounted() ? java_thread->vthread() : nullptr;
+  if (mounted_vt != nullptr && !JvmtiEnvBase::is_vthread_alive(mounted_vt)) {
+    mounted_vt = nullptr;
+  }
   assert(threadObj != nullptr, "sanity check");
 
-  // only need to get the thread's tag once per thread
-  jlong thread_tag = tag_for(_tag_map, threadObj);
+  StackRootCollector stack_collector(tag_map(), blk, java_thread);
 
-  // also need the thread id
-  jlong tid = java_lang_Thread::thread_id(threadObj);
+  if (!java_thread->has_last_Java_frame()) {
+    // this may be only platform thread
+    assert(mounted_vt == nullptr, "must be");
```

Comment:
I'm not sure this assert is right.
I think, a virtual thread may have an empty stack observable from a VM_op,
for instance when it is in a process of being terminated.
Though, it is not that easy to make this assert fired with a test case and prove this can happen.
Another danger is that a virtual thread can be observed from a VM_op as in a VTMS (mount/unmount) transition. I need to think a little bit about possible consequences. Is it better to treat current thread identity as of a carrier thread in such a case?

Reply:
removed the assert for safety.
I have no idea how vthread stack (frames on carrier thread and stack chunks) can look like during VTMS transitions (and it's very hard to reproduce the case by test)